lib/src/components/NxDropdown/AbstractDropdown.tsx

```diff
@@ -0,0 +1,152 @@
+    onCloseKeyDown: onCloseKeyDownProp,
+    ...attrs
+  } = props;
+
+  const menuRef = useRef<HTMLDivElement>(null);
+  const toggleRef = useRef<HTMLButtonElement>(null);
+
+  const onKeyDown: KeyboardEventHandler<HTMLDivElement> = event => {
+    if (isOpen && !disabled && event.key === 'Escape' && onToggleCollapseProp) {
+      if (onCloseKeyDownProp) {
+        onCloseKeyDownProp(event);
+      }
+
+      if (!event.defaultPrevented) {
+        onToggleCollapseProp();
+
+        // try to prevent the ESC from triggering multiple things, for instance closing a surrounding modal
+        // in addition to the dropdown
+        event.preventDefault();
+      }
+    }
+
+    if (onKeyDownProp) {
+      onKeyDownProp(event);
+    }
+  };
+
+  const handleDocumentClick = (event: MouseEvent) => {
+    if (isOpen && onToggleCollapseProp) {
+      let defaultPrevented = false;
+
+      if (onCloseClick) {
+        /* when consuming code calls preventDefault on the event, we only want that to affect this dropdown.
+        * Since the mouse clicks are captured at the document level, calling preventDefault on the actual
+        * native event could have a wide variety of other effects, so we proxy the event object overriding
+        * preventDefault with our own impl that only affects a locally-scoped variable
+        */
+        const proxiedClickEvent = new Proxy(event, {
+          get(target, prop) {
+            if (prop === 'preventDefault') {
+              return () => { defaultPrevented = true; };
+            }
+            else {
+              return Reflect.get(target, prop);
+            }
+          }
+        });
+
+        onCloseClick(proxiedClickEvent);
+      }
+
+      if (!defaultPrevented) {
+        onToggleCollapseProp();
+      }
+    }
+  };
+
+  function onToggleCollapse() {
+    // When closing, this is handled by handlDocumentClick. So only do it here when opening
+    // (i.e. when not already open)
+    !isOpen && onToggleCollapseProp && onToggleCollapseProp();
+  }
+
+  useEffect(() => {
+    if (disabled) {
+      return;
+    }
+
+    document.addEventListener('click', handleDocumentClick);
+
+    return function cleanup() {
+      document.removeEventListener('click', handleDocumentClick);
+    };
+  }, [disabled, isOpen]);
+
+  // When the dropdown is closed while focus is within it, set focus back to the dropdown toggle. Otherwise
+  // it goes back to the <body> which is less helpful especially when within a modal
+  function onMenuClosing() {
+    /* eslint-disable @typescript-eslint/no-non-null-assertion */
+    const focusedEl = document.activeElement,
+        menuEl = menuRef.current,
+        toggleEl = toggleRef.current;
+
+    if (menuEl && menuEl.contains(focusedEl)) {
+      toggleEl!.focus();
+    }
+    /* eslint-enable @typescript-eslint/no-non-null-assertion */
+  }
+
+  return (
+    <div ref={ref}
+         className={className}
+         onKeyDown={onKeyDown}
+         {...attrs}
+    >
```

Comment:
This seems like it would fit on one line

Reply:
> This seems like it would fit on one line

Updated: 86fda305f98b1a5df940fdaeba47240a69561ec0